### ISSUE TITLE
CuDNN + H100 Gives Weird Gradients

### ISSUE THREAD

**Issue by alanhdu** [pytorch/pytorch#150700]
### 🐛 Describe the bug

If I run this reproduction script at  https://gist.github.com/alanhdu/68aeca1b1cfbe63fe3464541a201fb79 (using the [arr.txt](https://github.com/user-attachments/files/19609554/arr.txt) file in `/tmp/arr.pt`), then I see something quite strange.

On an A100 GPU , I get results like:

```
cpu bf16 True 1.5634121894836426 0.3220250606536865
cpu fp32 True 1.5633708238601685 0.32201677560806274
cuda bf16 False 1.56340754032135 0.32201358675956726
cuda fp32 False 1.5633708238601685 0.32201698422431946
cuda bf16 True 1.5634031295776367 0.3220202922821045
cuda fp32 True 1.5633710622787476 0.32200005650520325
```

This makes sense to me -- while there are minor differences in the loss and the gradient norm, all of them are quite small and approximately the same.

But on an H100 GPU, I get results like:
```
cpu bf16 True 1.5634121894836426 0.3220250606536865
cpu fp32 True 1.5633708238601685 0.32201677560806274
cuda bf16 False 1.563407301902771 0.32201361656188965
cuda fp32 False 1.5633708238601685 0.32201698422431946
cuda bf16 True 1.5724581480026245 0.08233493566513062
cuda fp32 True 1.5720577239990234 0.18226394057273865
```

This is much more confusing -- for some reason, even though the loss function is roughly the same, the gradient norms are *very* different when using the CuDNN backend!

This is a very shrunken down version of a model internally where the model trains just fine on an A100 but fails to converge at all on an H100:
![Image](https://github.com/user-attachments/assets/f723bdaf-b6d5-4d8b-b530-0ab093f8b68d).

### Versions

This is running on the internal Meta cluster, so we are using a nightly version of PyTorch and Python 3.12. I've attached the collect env information at ([A100](https://gist.github.com/alanhdu/ccbd54e378a44159a0e39031fbac0ac4) and [H100](https://gist.github.com/alanhdu/bf43069780f4859c4aa96d3241a27efd)), although.

On both nodes, I get:
```
torch.version.cuda='12.4.0'
torch.backends.cudnn.version()=8903
```

**Comment by ngimel** [comment]
cc @eqy

**Comment by alanhdu** [comment]
I reran this again with CuDNN 9.2 (`torch.version.cuda=12.4.0, torch.backends.cudnn.version()=90200`).

On the H100, I get:

```
cpu bf16 True 1.5634121894836426 0.3220250606536865
cpu fp32 True 1.5633708238601685 0.32201677560806274
cuda bf16 False 1.563407301902771 0.32201361656188965
cuda fp32 False 1.5633708238601685 0.32201698422431946
cuda bf16 True 1.5634034872055054 0.32201576232910156
cuda fp32 True 1.5724433660507202 0.18314369022846222
```

On the A100, I get:

```
cpu bf16 True 1.5634121894836426 0.3220250606536865
cpu fp32 True 1.5633708238601685 0.32201677560806274
cuda bf16 False 1.56340754032135 0.32201358675956726
cuda fp32 False 1.5633708238601685 0.32201698422431946
cuda bf16 True 1.5634031295776367 0.322020560503006
cuda fp32 True 1.5633691549301147 0.32201123237609863
```

So something still seems to be wrong on the the H100.

**Comment by janeyx99** [comment]
Could you let us know what version of torch you're running on by running the collect_env script? https://github.com/pytorch/pytorch/blob/861d2cc02cce860d789cfda644a366abb95b53a5/torch/utils/collect_env.py#L4

**Comment by eqy** [comment]
Could you also please the shape of the data `X` loaded from `arr.pt` here? Ideally I would like to avoid loading arbitrary binary data from a blob.

**Comment by alanhdu** [comment]
No worries -- it's a `[16, 14, 16_000]` tensor.

**Comment by eqy** [comment]
I can reproduce the issue on my end but it looks like it's been fixed as of cuDNN 9.5.1+ where the same workload appears to dispatch to different kernels. Are you able to upgrade to e.g., the 12.6 wheel which would have a newer version of cuDNN?

**Comment by alanhdu** [comment]
We have a build with Cuda 12.8 + CuDNN 9.7 which does seem to solve the issue for this example! I will try re-run our full training loop to see if that helps things.

**Comment by alanhdu** [comment]
@eqy Yeah, CuDNN 9.7 seems to fix the issue. Do you have any information on what the original issue was? I'm wondering whether there are any experiments we should re-run on top of CuDNN 9.7, so knowing what the issue was would be super helopful!

**Comment by eqy** [comment]
My current assessment is that there are some problematic kernels which are no longer selected in 9.5.1+ for this workload---that has already been forwarded to the cuDNN team.
